lib/pathfinder/PathfinderOptions.cpp

```diff
@@ -44,6 +45,18 @@ PathfinderConfig::PathfinderConfig(std::shared_ptr<INodeStorage> nodeStorage, st
 {
 }
 
+void PathfinderConfig::checkIfTownlessAndAllowOneWayTeleports(const CGHeroInstance * hero)
+{
+	// FIXME: should be enabled by default with some bigger AI refactor so that main heroes are strong enough to
+	// go through and/or enable scouts (at some point ?) to visit such monoliths
+	// see first TODO in the top of PathfinderOptions.h
+	if (hero->cb->howManyTowns(hero->getOwner()) == 0)
+	{
+		options.useTeleportOneWay = true;
+		options.useTeleportOneWayRandom = true;
+	}
+}
+
```

Comment:
1) Should be in AI code. I believe AI configures its pathfinder parameters somewhere. Should be somewhere in there
2) This would not detect case if this particular hero is in small room with one-way teleport BUT has town someplace else